akka-persistence/src/main/scala/akka/persistence/Persistence.scala

```diff
@@ -87,51 +113,72 @@ object Persistence extends ExtensionId[Persistence] with ExtensionIdProvider {
  * Persistence extension.
  */
 class Persistence(val system: ExtendedActorSystem) extends Extension {
+  import scala.collection.concurrent.TrieMap
+
+  private def log: LoggingAdapter = Logging(system, getClass.getName)
+
   private val DefaultPluginDispatcherId = "akka.persistence.dispatchers.default-plugin-dispatcher"
+
   private val config = system.settings.config.getConfig("akka.persistence")
 
+  private val defaultJournalPluginId = config.getString("journal.plugin")
+
+  private val defaultSnapshotPluginId = config.getString("snapshot-store.plugin")
+
   val settings = new PersistenceSettings(config)
 
-  private val snapshotStore = createPlugin("snapshot-store") { _ ⇒
+  private def journalDispatchSelector(klaz: Class[_]): String =
+    if (classOf[AsyncWriteJournal].isAssignableFrom(klaz)) Dispatchers.DefaultDispatcherId else DefaultPluginDispatcherId
+
+  private def snapshotDispatchSelector(klaz: Class[_]): String =
     DefaultPluginDispatcherId
-  }
 
-  private val journal = createPlugin("journal") { clazz ⇒
-    if (classOf[AsyncWriteJournal].isAssignableFrom(clazz)) Dispatchers.DefaultDispatcherId
-    else DefaultPluginDispatcherId
-  }
+  /** Discovered persistence journal plugins - by journal plugin id. */
+  private val journalPlugins = TrieMap[String, ActorRef]()
 
-  /**
-   * Creates a canonical persistent actor id from a persistent actor ref.
-   */
-  def persistenceId(persistentActor: ActorRef): String = id(persistentActor)
+  /** Discovered persistence snapshot store plugins - by snapshot plugin id. */
+  private val snapshotPlugins = TrieMap[String, ActorRef]()
+
+  /** Check for default identity. */
+  private def isDefault(text: String) = text == null || text.length == 0
 
   /**
-   * Returns a snapshot store for a persistent actor identified by `persistenceId`.
+   * Returns a journal plugin actor identified by `journalPluginId`.
+   * When empty, looks in `akka.persistence.journal.plugin` to find configuration entry path.
+   * When configured, uses `journalPluginId` as absolute path to the journal configuration entry.
+   * Configuration entry must contain few required fields, such as `class`. See `src/main/resources/reference.conf`.
    */
-  def snapshotStoreFor(persistenceId: String): ActorRef = {
-    // Currently returns a snapshot store singleton but this methods allows for later
-    // optimizations where each persistent actor can have its own snapshot store actor.
-    snapshotStore
+  def journalFor(journalPluginId: String): ActorRef = {
+    val configPath = if (isDefault(journalPluginId)) defaultJournalPluginId else journalPluginId
+    // TODO remove synchronized after https://issues.scala-lang.org/browse/SI-7943
+    journalPlugins.synchronized(journalPlugins.getOrElseUpdate(configPath, createPlugin(configPath)(journalDispatchSelector)))
```

Comment:
repeating my previous comment about this, so that other reviewers can think about it:

`synchronized` might not be an acceptable solution, and I'm not sure we can rely on SI-7943.

That is a difficult concurrency problem. The tricky part is to only invoke the creator function once.

We have solved that in the registration of Akka extensions: https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/ActorSystem.scala#L745
That is probably not something we would like to duplicate.

It would be possible to register each journal as an extension, but that would require a `ConcurrentHashMap[String, ExtensionId] here for the mapping between the journalPluginId and ExtensionId instance.
I'm not sure if that is the best solution.